File: Java/803.java
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {

    static final int HIT = -1;
    static final int BRICK = 1;
    static final int VISITED = 2;
	
    public int[] hitBricks(int[][] grid, int[][] hits) {
        int [] result = new int[hits.length];
		
		// Mark erased bricks
        for (int [] hit : hits) {
            if (grid[hit[0]][hit[1]] == BRICK) {
                grid[hit[0]][hit[1]] = HIT;
            }
        }
		/*
			Run Depth First Search from top row in order to know 
			which bricks are left after every erase operation.
			We should mark these bricks as visited.
		*/
        for (int j = 0; j < grid[0].length; j++) {
            if (grid[0][j] == BRICK) {
                dfs(grid, 0, j);
            }
        }
		/*
			Run the hits array in reverse order.
			We should run a new DFS from here in order to know which cells can be visited.
			Count the new visited cells except the DFS root cell (because it was erased, it didn't fall)
		*/
        for (int k = result.length - 1; k >= 0; k--) {
            int i = hits[k][0], j = hits[k][1];
            if (grid[i][j] == HIT) {
                grid[i][j] = BRICK;
                if (isConnected(grid, i, j)) result[k] = dfs(grid, i, j) - 1;
            }
        }
        return result;
    }
	
    public boolean isConnected(int [][] grid, int i, int j) {
        return i - 1 < 0 || (grid[i - 1][j] == VISITED) ||
                (i + 1 < grid.length && grid[i + 1][j] == VISITED) ||
                (j - 1 >= 0 && grid[i][j - 1] == VISITED) ||
                (j + 1 < grid[0].length && grid[i][j + 1] == VISITED);
    }
	
    public int dfs(int [][] grid, int i, int j) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] != BRICK) {
            return 0;
        }
        grid[i][j] = VISITED;
        int sum = 1;
        sum += dfs(grid, i + 1, j);
        sum += dfs(grid,  i - 1, j);
        sum += dfs(grid, i, j + 1);
        sum += dfs(grid, i, j - 1);
        return sum;
    }
}
__________________________________________________________________________________________________
sample 51148 kb submission
class Solution {
    public int[] hitBricks(int[][] grid, int[][] hits) {
        int R = grid.length, C = grid[0].length;
        int[] dr = {1, 0, -1, 0};
        int[] dc = {0, 1, 0, -1};

        int[][] A = new int[R][C];
        for (int r = 0; r < R; ++r)
            A[r] = grid[r].clone();
        for (int[] hit: hits)
            A[hit[0]][hit[1]] = 0;

        DSU dsu = new DSU(R*C + 1);
        for (int r = 0; r < R; ++r) {
            for (int c = 0; c < C; ++c) {
                if (A[r][c] == 1) {
                    int i = r * C + c;
                    if (r == 0)
                        dsu.union(i, R*C);
                    if (r > 0 && A[r-1][c] == 1)
                        dsu.union(i, (r-1) *C + c);
                    if (c > 0 && A[r][c-1] == 1)
                        dsu.union(i, r * C + c-1);
                }
            }
        }
        int t = hits.length;
        int[] ans = new int[t--];

        while (t >= 0) {
            int r = hits[t][0];
            int c = hits[t][1];
            int preRoof = dsu.top();
            if (grid[r][c] == 0) {
                t--;
            } else {
                int i = r * C + c;
                for (int k = 0; k < 4; ++k) {
                    int nr = r + dr[k];
                    int nc = c + dc[k];
                    if (0 <= nr && nr < R && 0 <= nc && nc < C && A[nr][nc] == 1)
                        dsu.union(i, nr * C + nc);
                }
                if (r == 0)
                    dsu.union(i, R*C);
                A[r][c] = 1;
                ans[t--] = Math.max(0, dsu.top() - preRoof - 1);
            }
        }

        return ans;
    }
}

class DSU {
    int[] parent;
    int[] rank;
    int[] sz;

    public DSU(int N) {
        parent = new int[N];
        for (int i = 0; i < N; ++i)
            parent[i] = i;
        rank = new int[N];
        sz = new int[N];
        Arrays.fill(sz, 1);
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public void union(int x, int y) {
        int xr = find(x), yr = find(y);
        if (xr == yr) return;

        if (rank[xr] < rank[yr]) {
            int tmp = yr;
            yr = xr;
            xr = tmp;
        }
        if (rank[xr] == rank[yr])
            rank[xr]++;

        parent[yr] = xr;
        sz[xr] += sz[yr];
    }

    public int size(int x) {
        return sz[find(x)];
    }

    public int top() {
        return size(sz.length - 1) - 1;
    }
}
__________________________________________________________________________________________________

File: Python3/803.py
__________________________________________________________________________________________________
sample 604 ms submission
class Solution:
    def hitBricks(self, grid: List[List[int]], hits: List[List[int]]) -> List[int]:
        if not grid or not grid[0]:
            return []
        
        m, n = len(grid), len(grid[0])
        EMPTY = 0
        BRICK = 1
        DROP = 2
        WAIT = 3
        
        def count(x, y):
            cnt = 1
            grid[x][y] = DROP
            
            if x > 0 and grid[x - 1][y] == BRICK:
                cnt += count(x - 1, y)
            if y > 0 and grid[x][y - 1] == BRICK:
                cnt += count(x, y - 1)
            if x + 1 < m and grid[x + 1][y] == BRICK:
                cnt += count(x + 1, y)
            if y + 1 < n and grid[x][y + 1] == BRICK:
                cnt += count(x, y + 1)
            
            return cnt
        
        for x, y in hits:
            if grid[x][y] == BRICK:
                grid[x][y] = WAIT
        
        for y in range(n):
            if grid[0][y] == BRICK:
                count(0, y)
        
        ans = []
        
        for x, y in hits[::-1]:
            if grid[x][y] == WAIT:
                grid[x][y] = BRICK
                
                if (
                    x == 0 or
                    (x > 0 and grid[x - 1][y] == DROP) or 
                    (y > 0 and grid[x][y - 1] == DROP) or 
                    (x + 1 < m and grid[x + 1][y] == DROP) or 
                    (y + 1 < n and grid[x][y + 1] == DROP)
                ):
                    ans.append(count(x, y) - 1)
                else:
                    ans.append(0)
            else:
                ans.append(0)
        
        return ans[::-1]
__________________________________________________________________________________________________
sample 608 ms submission
class Solution:
    """
    T: O(m * n)
    S: O(m * n)
    """
    def hitBricks(self, grid: List[List[int]], hits: List[List[int]]) -> List[int]:
        if not grid or not grid[0] or not hits:
            return []
        
        m, n, k = len(grid), len(grid[0]), len(hits)
        EMPTY = 0
        BRICK = 1
        ERASURE = 2
        DROPED = 3
        
        def count(x, y):
            grid[x][y] = DROPED
            cnt = 1
            
            if x > 0 and grid[x - 1][y] == BRICK:
                cnt += count(x - 1, y)
            if y > 0 and grid[x][y - 1] == BRICK:
                cnt += count(x, y - 1)
            if x + 1 < m and grid[x + 1][y] == BRICK:
                cnt += count(x + 1, y)
            if y + 1 < n and grid[x][y + 1] == BRICK:
                cnt += count(x, y + 1)
            
            return cnt
        
        for x, y in hits:
            if grid[x][y] == BRICK:
                grid[x][y] = ERASURE
        
        for y in range(n):
            if grid[0][y] == BRICK:
                count(0, y)
        
        ans = [0] * k
        
        for i in range(k - 1, -1, -1):
            x, y = hits[i]
            
            if grid[x][y] == ERASURE:
                grid[x][y] = BRICK
                
                if (
                    (x == 0) or
                    (x > 0 and grid[x - 1][y] == DROPED) or 
                    (y > 0 and grid[x][y - 1] == DROPED) or
                    (x + 1 < m and grid[x + 1][y] == DROPED) or
                    (y + 1 < n and grid[x][y + 1] == DROPED)
                ):
                    ans[i] = count(x, y) - 1
        
        return ans
__________________________________________________________________________________________________
